langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java

```diff
@@ -166,11 +169,34 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Exceptio
                             return new AiServiceTokenStream(messages, contents, context, memoryId); // TODO moderation
                         }
 
+
+                        // Begin tools
+                        List<ToolSpecification> toolSpecifications = context.toolSpecifications;
+                        Map<String, ToolExecutor> toolExecutors = context.toolExecutors;
+
+                        if (context.toolProvider != null) {
+                            // Use provider to decide matching tools
+                            ChatMemory chatMemory = context.hasChatMemory()
+                                    ? context.chatMemory(memoryId)
+                                    : null;
+                            toolSpecifications = new ArrayList<>();
+                            toolExecutors = new HashMap<>();
+                            ToolProviderRequest providerRequest = new ToolProviderRequest(chatMemory, userMessage);
```

Comment:
```suggestion
                            ToolProviderRequest providerRequest = new ToolProviderRequest(memoryId, userMessage);
```

`chatMemory` above can be removed